packages/sources/stader-balance/src/endpoint/balance.ts

```diff
@@ -0,0 +1,741 @@
+            // Get inactive pool balance from Stader's StakePoolManager contract
+            this.getStaderStakeManagerBalance(req, blockTag, balances),
+            // Get balance of all execution layer reward addresses
+            this.getElRewardBalances(req, blockTag, balances, context.adapterSettings),
+            // Get balance of the Permissioned Pool address
+            this.getPermissionedPoolBalance(req, blockTag, balances),
+          ])
+
+          await Promise.all([
+            // Perform validator level calculations
+            await this.performValidatorCalculations(
+              addresses,
+              validatorStateList,
+              balances,
+              req,
+              validatorDeposit,
+              ethDepositContract,
+              commissionMap,
+              collateralEthMap,
+              blockTag,
+            ),
+            // Get permissionless/permissioned pool address balances
+            this.getPoolAddressBalances(
+              req,
+              blockTag,
+              balances,
+              validatorDeposit,
+              commissionMap,
+              collateralEthMap,
+            ),
+          ])
+
+          response = {
+            data: {
+              result: balances,
+            },
+            result: null,
+            statusCode: 200,
+            timestamps: {
+              providerDataRequestedUnixMs,
+              providerDataReceivedUnixMs: Date.now(),
+              providerIndicatedTimeUnixMs: undefined,
+            },
+          }
+        } catch (e) {
+          const errorMessage = e instanceof Error ? e.message : 'Unknown error occurred'
+          logger.error(errorMessage)
+          response = buildErrorResponse(errorMessage, providerDataRequestedUnixMs)
+        }
+
+        await this.responseCache.write(this.name, [
+          {
+            params: req,
+            response,
+          },
+        ])
+      }),
+    )
+  }
+
+  // Retrieve balances from the beacon chain for all validators in request
+  async queryValidatorBalances(
+    req: RequestParams,
+    settings: typeof config.settings,
+  ): Promise<ValidatorState[]> {
+    try {
+      const url = `/eth/v1/beacon/states/${req.stateId}/validators`
+      const statusList = req.validatorStatus?.join(',')
+      const batchSize = settings.BATCH_SIZE
+      const batchedAddresses = []
+      const responses: AxiosResponse<ProviderResponse>[] = []
+      const result: ValidatorState[] = []
+      const addresses = req.addresses
+      // Separate the address set into the specified batch size
+      // Add the batches as comma-separated lists to a new list used to make the requests
+      for (let i = 0; i < addresses.length / batchSize; i++) {
+        batchedAddresses.push(
+          addresses
+            .slice(i * batchSize, i * batchSize + batchSize)
+            .map(({ address }) => address)
+            .join(','),
+        )
+      }
```

Comment:
same thing as above, all these batching steps should be abstracted into a common pattern to make it easier to read